packages/material-ui-styles/src/withStyles.js

```diff
@@ -53,22 +56,56 @@ const withStyles = (stylesOrCreator, options = {}) => Component => {
     if (typeof name === 'string' || withTheme) {
       // name and withTheme are invariant in the outer scope
       // eslint-disable-next-line react-hooks/rules-of-hooks
-      theme = useTheme();
+      theme = useTheme() || defaultTheme;
 
       if (name) {
         more = getThemeProps({ theme, name, props: other });
       }
 
       // Provide the theme to the wrapped component.
       // So we don't have to use the `withTheme()` Higher-order Component.
-      if (withTheme) {
+      if (withTheme && !more.theme) {
         more.theme = theme;
       }
     }
 
     return <Component ref={innerRef || ref} classes={classes} {...more} />;
   });
 
+  if (process.env.NODE_ENV === 'test' && !ponyfillGlobal.disableShallowSupport) {
```

Comment:
What's this for?

Reply:
This removes the need for migrating all the tests away from the shallow API. We can consider killing this code branch once we have removed all the shallow tests. It has a second advantage, it's making the class names deterministic in the test environment. Related to #14358.

Reply:
>it makes the class names deterministic in the test environment

That is a lot of redundant code for behavior that is only required for a single testing pattern that is already solvable. Emotion and styled-components agree with me here. I hope this gets removed once we figure out shallow testing.

Reply:
Yes, I think that this whole code branch should get removed.